docs/closure-compiler.md

```diff
@@ -1,179 +0,0 @@
-
-### Standard foundation constants must be defined as `enum` types
-
-- `cssClasses` and `strings` should be [string enums](https://www.typescriptlang.org/docs/handbook/enums.html#string-enums)
-- `numbers` should be a [numeric enum](https://www.typescriptlang.org/docs/handbook/enums.html#numeric-enums)
-
-```ts
-enum cssClasses {
-  // ...
-}
-
-enum strings {
-  // ...
-}
-
-enum numbers {
-  // ...
-}
-
-export {cssClasses, strings, numbers};
-```
-
-### All adapters must be defined as interfaces
-
-Each adapter must be defined within an `adapter.ts` file in the component's package directory.
-All methods should contain a summary of what they should do. This summary should be
-copied over to the adapter API documentation in our README. This will facilitate future endeavors
-to potentially automate the generation of our adapter API docs. _Note that this replaces the
-inline comments present in the methods within `defaultAdapter`_.
-
-```ts
-// adapter.ts
-
-interface MDCComponentAdapter {
-  /**
-   * Adds a class to the root element.
-   */
-  addClass(className: string): void;
-
-  /**
-   * Removes a class from the root element.
-   */
-  removeClass(className: string): void;
-}
-
-export default MDCComponentAdapter;
-```
-
-### Foundation classes must extend `MDCFoundation`
-
-Foundations must extend `MDCFoundation` parameterized by their respective adapter. The
-`defaultAdapter` must return an object with the correct adapter shape.
-
-```ts
-// foundation.ts
-
-import {MDCFoundation} from '@material/base/foundation';
-import MDCComponentAdapter from './adapter';
-
-class MDCComponentFoundation extends MDCFoundation<MDCComponentAdapter> {
-  static get defaultAdapter(): MDCComponentAdapter {
-    return {
-      addClass: (className: string) => {},
-      removeClass: (className: string) => {},
-    };
-  }
-}
-
-export default MDCComponentFoundation;
-```
-
-### Component classes must extend `MDCComponent`
-
-Components must extend `MDCComponent` parameterized by their respective foundation.
-
-```ts
-// index.ts
-
-import {MDCComponent} from '@material/base/component';
-import MDCComponentFoundation from './foundation';
-
-class MDCAwesomeComponent extends MDCComponent<MDCComponentFoundation> {
-  getDefaultFoundation(): MDCComponentFoundation {
-    return new MDCComponentFoundation({
-      addClass: (className: string) => this.root_.classList.add(className),
-      removeClass: (className: string) => this.root_.classList.remove(className),
-    });
-  }
-}
-
-export default MDCAwesomeComponent;
-```
-
-### Interfaces for type definitions always end in `Type`
-
-<!--{% raw %} -->
-```ts
-interface EventDataType {
-  foo: string;
-  bar: number;
-}
-```
-<!--{% endraw %} -->
-
-Using this convention allows us to write tooling around handling these expressions, such as lint rule exceptions.
-
-### For key/value maps, use index signatures where possible
```

Comment:
This should be moved to best_practices.md; I added this specifically for TypeScript

Reply:
Re both closure-compiler comments: I ended up adding this entire block:
```

#### All `import` statements must _not_ use re-exported modules

Only the `index.ts` or `component.ts` files are allowed to reference from other component packages' `index.ts`. This is because wrapping libraries only use `foundation` and `adapter`, so we should decouple the `component`.

```ts
// BAD
import {MDCFoundation} from '@material/base';
// GOOD
import {MDCFoundation} from '@material/base/foundation';
```

#### All adapters must be defined as interfaces

Each adapter must be defined within an `adapter.ts` file in the component's package directory.
All methods should contain a summary of what they should do. This summary should be
copied over to the adapter API documentation in our README. This will facilitate future endeavors
to potentially automate the generation of our adapter API docs. _Note that this replaces the
inline comments present in the methods within `defaultAdapter`_.

 ```ts
// adapter.ts
export interface MDCComponentAdapter {
  /**
   * Adds a class to the root element.
   */
  addClass(className: string): void;
   /**
   * Removes a class from the root element.
   */
  removeClass(className: string): void;
}
```

#### Foundation classes must extend `MDCFoundation`

 Foundations must extend `MDCFoundation` parameterized by their respective adapter. The
`defaultAdapter` must return an object with the correct adapter shape.

```ts
// foundation.ts
import {MDCFoundation} from '@material/base/foundation';
import MDCComponentAdapter from './adapter';
export class MDCComponentFoundation extends MDCFoundation<MDCComponentAdapter> {
  static get defaultAdapter(): MDCComponentAdapter {
    return {
      addClass: (className: string) => undefined,
      removeClass: (className: string) => undefined,
    };
  }
}
```

#### Component classes must extend `MDCComponent`

Components must extend `MDCComponent` parameterized by their respective foundation.

```ts
// index.ts
import {MDCComponent} from '@material/base/component';
import MDCComponentFoundation from './foundation';
export class MDCAwesomeComponent extends MDCComponent<MDCComponentFoundation> {
  getDefaultFoundation(): MDCComponentFoundation {
    return new MDCComponentFoundation({
      addClass: (className: string) => this.root_.classList.add(className),
      removeClass: (className: string) => this.root_.classList.remove(className),
    });
  }
}
```

### For key/value maps, use index signatures where possible

Index signatures are useful for homogeneous maps of key/value pairs, while interfaces are useful for specific object signatures.
Both are more specific than just typing something as `object`, and are thus preferable for their respective use cases.

```ts
static get strings(): {[key: string]: string} {
  ...
}
```